db/pgx/sql.go

```diff
@@ -0,0 +1,263 @@
+	`ON CONFLICT (mt_id, key) DO UPDATE SET type = $3, child_l = $4, child_r = $5, entry = $6`
+
+const updateRootStmt = `INSERT INTO mt_roots (mt_id, key) VALUES ($1, $2) ` +
+	`ON CONFLICT (mt_id) DO UPDATE SET key = $2`
+
+type DB interface {
+	Exec(ctx context.Context, sql string, arguments ...interface{}) (pgconn.CommandTag, error)
+	Query(ctx context.Context, sql string, args ...interface{}) (pgx.Rows, error)
+	QueryRow(ctx context.Context, sql string, args ...interface{}) pgx.Row
+}
+
+// Storage implements the db.Storage interface
+type Storage struct {
+	db             DB
+	mtId           uint64
+	currentVersion uint64
+	currentRoot    *merkletree.Hash
+}
+
+type NodeItem struct {
+	MTId uint64 `db:"mt_id"`
+	Key  []byte `db:"key"`
+	// Type is the type of node in the tree.
+	Type byte `db:"type"`
+	// ChildL is the left child of a middle node.
+	ChildL []byte `db:"child_l"`
+	// ChildR is the right child of a middle node.
+	ChildR []byte `db:"child_r"`
+	// Entry is the data stored in a leaf node.
+	Entry     []byte  `db:"entry"`
+	CreatedAt *uint64 `db:"created_at"`
+	DeletedAt *uint64 `db:"deleted_at"`
+}
+
+type RootItem struct {
+	MTId      uint64  `db:"mt_id"`
+	Key       []byte  `db:"key"`
+	CreatedAt *uint64 `db:"created_at"`
+	DeletedAt *uint64 `db:"deleted_at"`
+}
+
+// NewSqlStorage returns a new Storage
+func NewSqlStorage(db DB, mtId uint64) *Storage {
+	return &Storage{db: db, mtId: mtId}
+}
+
+// WithPrefix implements the method WithPrefix of the interface db.Storage
+func (s *Storage) WithPrefix(prefix []byte) merkletree.Storage {
+	//return &Storage{db: s.db, prefix: merkletree.Concat(s.prefix, prefix)}
+	// TODO: remove WithPrefix method
+	mtId, _ := binary.Uvarint(prefix)
+	return &Storage{db: s.db, mtId: mtId}
+}
+
+// Get retrieves a value from a key in the db.Storage
+func (s *Storage) Get(ctx context.Context,
+	key []byte) (*merkletree.Node, error) {
+	item := NodeItem{}
+	row := s.db.QueryRow(ctx, `SELECT mt_id, key, type, child_l, child_r, entry, created_at, deleted_at
+		FROM mt_nodes WHERE mt_id = $1 AND key = $2`, s.mtId, key)
+	err := row.Scan(&item.MTId,
+		&item.Key,
+		&item.Type,
+		&item.ChildL,
+		&item.ChildR,
+		&item.Entry,
+		&item.CreatedAt,
+		&item.DeletedAt)
+	if err != nil && errors.Is(err, pgx.ErrNoRows) {
+		return nil, merkletree.ErrNotFound
+	} else if err != nil {
+		return nil, err
+	}
+
+	node, err := item.Node()
+	if err != nil {
+		return nil, err
+	}
+	return node, nil
+}
+
+func (s *Storage) Put(ctx context.Context, key []byte,
+	node *merkletree.Node) error {
+
+	var childL []byte
+	if node.ChildL != nil {
+		childL = append(childL, node.ChildL[:]...)
+	}
+
+	var childR []byte
+	if node.ChildR != nil {
+		childR = append(childR, node.ChildR[:]...)
+	}
+
+	var entry []byte
+	if node.Entry[0] != nil && node.Entry[1] != nil {
+		entry = append(node.Entry[0][:], node.Entry[1][:]...)
+	}
+
+	_, err := s.db.Exec(ctx, upsertStmt, s.mtId, key[:], node.Type,
```

Comment:
We can't do this. Since `node.ChildL` and `node.ChildR` can be `nil`
We can do next:
```go
if node.ChildR == nil {
      node.ChildR = []byte{}
}
```
But for me, it is no make sense.